lib/ui/fake_flutter/print.dart

```diff
@@ -0,0 +1,184 @@
+
+part of 'fake_flutter.dart';
+
+/// Signature for [debugPrint] implementations.
+typedef DebugPrintCallback = void Function(String message, {int wrapWidth});
+
+/// Prints a message to the console, which you can access using the "flutter"
+/// tool's "logs" command ("flutter logs").
+///
+/// If a wrapWidth is provided, each line of the message is word-wrapped to that
+/// width. (Lines may be separated by newline characters, as in '\n'.)
+///
+/// By default, this function very crudely attempts to throttle the rate at
+/// which messages are sent to avoid data loss on Android. This means that
+/// interleaving calls to this function (directly or indirectly via, e.g.,
+/// [debugDumpRenderTree] or [debugDumpApp]) and to the Dart [print] method can
+/// result in out-of-order messages in the logs.
+///
+/// The implementation of this function can be replaced by setting the
+/// [debugPrint] variable to a new implementation that matches the
+/// [DebugPrintCallback] signature. For example, flutter_test does this.
+///
+/// The default value is [debugPrintThrottled]. For a version that acts
+/// identically but does not throttle, use [debugPrintSynchronously].
+DebugPrintCallback debugPrint = debugPrintThrottled;
+
+/// Alternative implementation of [debugPrint] that does not throttle.
+/// Used by tests.
+void debugPrintSynchronously(String message, {int wrapWidth}) {
+  if (wrapWidth != null) {
+    print(message
+        .split('\n')
+        .expand<String>((String line) => debugWordWrap(line, wrapWidth))
+        .join('\n'));
+  } else {
+    print(message);
+  }
+}
+
+/// Implementation of [debugPrint] that throttles messages. This avoids dropping
+/// messages on platforms that rate-limit their logging (for example, Android).
+void debugPrintThrottled(String message, {int wrapWidth}) {
+  final List<String> messageLines = message?.split('\n') ?? <String>['null'];
+  if (wrapWidth != null) {
+    _debugPrintBuffer.addAll(messageLines
+        .expand<String>((String line) => debugWordWrap(line, wrapWidth)));
+  } else {
+    _debugPrintBuffer.addAll(messageLines);
+  }
+  if (!_debugPrintScheduled) _debugPrintTask();
+}
+
+int _debugPrintedCharacters = 0;
+const int _kDebugPrintCapacity = 12 * 1024;
+const Duration _kDebugPrintPauseTime = Duration(seconds: 1);
+final Queue<String> _debugPrintBuffer = Queue<String>();
+final Stopwatch _debugPrintStopwatch = Stopwatch();
+Completer<void> _debugPrintCompleter;
+bool _debugPrintScheduled = false;
+void _debugPrintTask() {
+  _debugPrintScheduled = false;
+  if (_debugPrintStopwatch.elapsed > _kDebugPrintPauseTime) {
+    _debugPrintStopwatch.stop();
+    _debugPrintStopwatch.reset();
+    _debugPrintedCharacters = 0;
+  }
+  while (_debugPrintedCharacters < _kDebugPrintCapacity &&
+      _debugPrintBuffer.isNotEmpty) {
+    final String line = _debugPrintBuffer.removeFirst();
+    _debugPrintedCharacters +=
+        line.length; // TODO(ianh): Use the UTF-8 byte length instead
```

Comment:
Probably remove this todo for ianh@?

Reply:
I think there is value for us just porting files wholesale instead of trying to cleanup as these files will be ported more cleanly at some later point for hummingbird. all we want for them now is that we can reasonably bring in new versions from package:flutter without odd diffs.